File: cpp/1713_得到子序列的最少操作次数_hard.cpp
class Solution {
public:
    int minOperations(vector<int>& target, vector<int>& arr) {
        int tLength = target.size();
        unordered_map <int, int> data;
        for (int i=0; i< tLength; i++)
            data[target[i]] = i;
        vector<int> ans;
        for (auto num : arr) {
            if (data.count(num)) {
                int n = ans.size();
                int value = data[num];
                if (n == 0 || value > ans[n-1])
                    ans.emplace_back(value);
                else {
                    int l = 0, r = n-1;
                    int pos = r;
                    while (l <= r) {
                        int mid = l + (r - l)/2;
                        if (ans[mid] < value)
                            l = mid + 1;
                        else {
                            pos = mid;
                            r = mid - 1;
                        }
                    }
                    ans[pos] = value;
                }
            }
        }
        return tLength - ans.size();
    }
};

File: python/1713_得到子序列的最少操作次数_hard.py
class Solution:
    def minOperations(self, target: List[int], arr: List[int]) -> int:
        t = len(target)
        tar = dict()
        for i in range(t):
            tar[target[i]] = i
        data = []
        for a in arr:
            if a in tar:
                data.append(tar[a])
        if not data:
            return t
        n = len(data)
        d = []
        for num in data:
            if not d or num > d[-1]:
                d.append(num)
            else:
                l, r = 0, len(d)-1
                pos = r
                while l <= r:
                    mid = l + (r - l) // 2
                    if d[mid] >= num:
                        pos = mid
                        r = mid - 1
                    else:
                        l = mid + 1
                d[pos] = num
        return t - len(d)